src/inference/graph_inference.cpp

```diff
@@ -786,6 +879,137 @@ class GraphNodeAssignment : public Nice2Assignment {
     }
   }
 
+  // Perform optimization based on factor features
+  void LocalPerFactorOptimizationPass(const GraphInference& fweights, size_t beam_size) {
+    std::vector<std::pair<double, Factor>> empty;
+    for (const Factor& factor : query_->factors_) {
+      std::vector<int> inf_vars;
+      Factor giv_labels;
+      for (auto var = factor.begin(); var != factor.end(); var++) {
+        Assignment a= assignments_[(*var)];
+        if (a.must_infer == true) {
+          inf_vars.push_back((*var));
+        } else {
+          giv_labels.insert(a.label);
+        }
+      }
+      if (inf_vars.size() <= 1) {
+        continue;
+      }
+
+      std::vector<Factor> factors;
+      GetFactorCandidates(fweights, factor.size(), &factors, giv_labels, beam_size);
+      double best_score = 0;
+      std::vector<int> best_assignments;
+      for (auto var = inf_vars.begin(); var != inf_vars.end(); var++) {
```

Comment:
This is a vector, sometimes you iterate it with iterator, sometimes with index. Be consistent, use index. Another trick is to preassign best_assignment with 0 and be as large as inf_vars, then set the values. This way you will avoid growing it with push_back. This is applicable to a few places in the code.